### PR TITLE
STY: migrate formatting from black to ruff-format

### LINKED ISSUE #4748: STY: switching from black to ruff-format ?
`ruff` now comes with a super fast formatter, which [draws heavily from black](https://docs.astral.sh/ruff/formatter/#black-compatibility). The formatter is currently in beta but is expected to reach maturity very soon.
There are [a couple intentional deviations from black's style](https://docs.astral.sh/ruff/formatter/black/), which currently affect about 100 lines of code in yt (roughly 0.1%), and to me they all make sense.

Do we want to migrate from black to ruff at some point ?

Expected gains:
- 3x faster runs for running pre-commit eagerly (`pre-commit run --all-files`) locally as well as pre-commit.ci (from 20s to 7s as of my latest measurement)
- one less pre-commit hook to keep updated and manage

Predictable costs:
- Currently there's no replacement for `blacken-docs` so we'd have to manage 2 very-slightly different styles and hooks if we migrated now
- contrary to black, ruff-format doesn't validate AST conservation at runtime (but its test suite does, see https://github.com/astral-sh/ruff/issues/8184), so it's marginally less safe than black.

To me gains seem to outweigh costs, so I'd propose to perform the migration sometimes after the formatter is out of beta, if a consensus is reached.

### PR DESCRIPTION
**Pull request by neutrinoceros**
## PR Summary

Closes #4748
The formatting commit is "semi-automated" because ruff was, in some places, somewhat akwardly moving around some strings formatted with `%`, but I found manually changing them to fstrings helped.

### CODE REVIEW

**Comment by neutrinoceros** [comment]
just a note on performance since it's the main selling point for ruff: `pre-commit.ci` run took 7.5s instead of the usual ~26s.

**Comment by neutrinoceros** [comment]
@yt-fido test this please

**Comment by neutrinoceros** [comment]
@yt-fido test this please

**Comment by neutrinoceros** [comment]
@matthewturk @cphyc @chrishavlin anyone up for reviewing this ?

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -27,21 +27,13 @@ repos:
     - id: check-executables-have-shebangs
     - id: check-yaml
 
--   repo: https://github.com/psf/black-pre-commit-mirror
-    rev: 24.3.0
-    hooks:
-    - id: black-jupyter
-
--   repo: https://github.com/adamchainz/blacken-docs
```
Comment (cphyc) [inline .pre-commit-config.yaml:35]: Sorry if I missed this information - but is ruff also running on docs? I couldn't find the confirmation online. The closest I got was https://github.com/astral-sh/ruff/issues/8237, but then I don't see where ruff is configured to run on the doc?

File: .git-blame-ignore-revs [hunk 2]
```diff
@@ -47,3 +47,6 @@ ec8bb45ea1603f3862041fa9e8ec274afd9bbbfd
 
 # auto upgrade typing idioms from Python 3.8 to 3.9
 4cfd370a8445abd4620e3853c2c047ee3d649fd7
+
+# migration: black -> ruff-format
+61ca693b2c58eefc0ca5768ee4c3a8512b47504a
```
[the remark this answers is not in the document]
Comment (neutrinoceros) [reply .git-blame-ignore-revs:52]: Indeed. Thanks for checking !

File: .pre-commit-config.yaml (hunk 1 above)
Comment (neutrinoceros) [reply .pre-commit-config.yaml:35]: hum. I may have tricked myself into believing that it was supported when I misunderstood something and didn't link the source in https://github.com/yt-project/yt/issues/4748#issuecomment-1971112987

I *think* the source of my confusion was https://github.com/astropy/astropy/pull/16132, which doesn't have anything to do with embedded python blocks.
I'll revert this and keep blacken-docs around for a little longer, but I'll also link the ruff issue. Thanks !

File: pyproject.toml [hunk 3]
```diff
@@ -243,27 +243,6 @@ zip-safe = false
...
-  | \.tox
-  | \.venv
-  | _build
-  | buck-out
-  | build
-  | dist
-  | yt/frontends/stream/sample_data
-)/
-| yt/visualization/_colormap_data.py
-'''
```
Comment (cphyc) [inline pyproject.toml:265]: Can we remove the black-specific configuration from `pyproject.toml` since it is still in use by blacken-docs?

Comment (neutrinoceros) [reply pyproject.toml]: Good point. I (mostly) reverted it, but I took the opportunity to clean up unnecessary parameters.

**Review by cphyc — APPROVED** [review]

**Comment by neutrinoceros** [comment]
I'm just going to merge now before pre-commit.ci auto PR create a merge conflict with this one. Thanks !

**Pull request merged**